src/java.desktop/windows/native/libawt/windows/awt_Frame.cpp

```diff
@@ -1577,63 +1580,65 @@ void AwtFrame::_NotifyModalBlocked(void *param)
+            HWND fHWnd = f->GetHWnd();
+            while (::GetParent(fHWnd) != NULL) {
+                fHWnd = ::GetParent(fHWnd);
+            }
+            // we must get a toplevel hwnd here, however due to some strange
+            // behaviour of Java Plugin (a bug?) when running in IE at
+            // this moment the embedded frame hasn't been placed into the
+            // browser yet and fHWnd is not a toplevel, so we shouldn't install
+            // the hook here
+            if ((::GetWindowLong(fHWnd, GWL_STYLE) & WS_CHILD) == 0) {
+                // if this toplevel is created in another thread, we should install
+                // the modal hook into it to track window activation and mouse events
+                DWORD fThread = ::GetWindowThreadProcessId(fHWnd, NULL);
+                if (fThread != AwtToolkit::GetInstance().MainThread()) {
+                    // check if this thread has been already blocked
+                    BlockedThreadStruct *blockedThread = (BlockedThreadStruct *)sm_BlockedThreads.get((void *)((intptr_t)fThread));
+                    if (blocked) {
+                        if (blockedThread == NULL) {
+                            blockedThread = new BlockedThreadStruct;
+                            blockedThread->framesCount = 1;
+                            blockedThread->modalHook = ::SetWindowsHookEx(WH_CBT, (HOOKPROC)AwtDialog::ModalFilterProc,
+                                                                          0, fThread);
+                            blockedThread->mouseHook = ::SetWindowsHookEx(WH_MOUSE, (HOOKPROC)AwtDialog::MouseHookProc_NonTT,
+                                                                          0, fThread);
+                            sm_BlockedThreads.put((void *)((intptr_t)fThread), blockedThread);
                         } else {
-                            blockedThread->framesCount--;
+                            blockedThread->framesCount++;
+                        }
+                    } else {
+                        // see the comment above: if Java Plugin behaviour when running in IE
+                        // was right, blockedThread would be always not NULL here
+                        if (blockedThread != NULL) {
+                            DASSERT(blockedThread->framesCount > 0);
+                            if ((blockedThread->framesCount) == 1) {
+                                ::UnhookWindowsHookEx(blockedThread->modalHook);
+                                ::UnhookWindowsHookEx(blockedThread->mouseHook);
+                                sm_BlockedThreads.remove((void *)((intptr_t)fThread));
+                                delete blockedThread;
+                            } else {
+                                blockedThread->framesCount--;
+                            }
                         }
                     }
                 }
             }
         }
     }
+
```

Comment:
A possible improvement later (and for a future RFE) would be to use RAII for deletion and then get rid of the labels. awt is one of the few places that uses C++ for native code, so why not.

Reply:
The comment seems answered. Please resolve this discussion.